gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java

```diff
@@ -0,0 +1,278 @@
+    Set<ContentKey> deletedKeys = new HashSet<>();
+    Consumer<LogResponse.LogEntry> perCommit =
+        logEntry -> handleLiveCommit(logEntry, deletedKeys, bloomFilterMap);
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            api, reference.getName(), null, null, null, OptionalInt.empty(), true)) {
+      Spliterator<LogResponse.LogEntry> src = commits.spliterator();
+      // Use a Spliterator to limit the processed commits to the "live" commits - i.e. stop at the
+      // first "non-live" commit.
+      new Spliterators.AbstractSpliterator<LogResponse.LogEntry>(src.estimateSize(), 0) {
+        private boolean more = true;
+        private boolean refHead = true;
+
+        @Override
+        public boolean tryAdvance(Consumer<? super LogResponse.LogEntry> action) {
+          if (!more) {
+            return false;
+          }
+          more =
+              src.tryAdvance(
+                  logEntry -> {
+                    if (!liveCommitPredicate.test(logEntry.getCommitMeta())) {
+                      // found first expired commit
+                      more = false;
+                    }
+                    // Process the commit if either the commit is considered live (i.e. the
+                    // liveCommitPredicate yields 'true') or at least for the very first commit
+                    // (aka the reference's HEAD).
+                    if (more || refHead) {
+                      action.accept(logEntry);
+                      refHead = false;
+                    }
+                  });
+          return more;
+        }
+      }.forEachRemaining(perCommit);
+    } catch (NessieNotFoundException e) {
+      LOG.info("Reference {} to retrieve commits for no longer exists", reference);
+    }
+    return bloomFilterMap;
+  }
+
+  private IdentifiedResult walkAllReference(
+      NessieApiV1 api,
+      Reference reference,
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap) {
+    final Logger LOG = LoggerFactory.getLogger(DistributedIdentifyUtil.class);
+    IdentifiedResult result = new IdentifiedResult();
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            api, "DETACHED", reference.getHash(), null, null, OptionalInt.empty(), true)) {
+      commits.forEach(logEntry -> handleCommit(logEntry, liveContentsBloomFilterMap, result));
+    } catch (NessieNotFoundException e) {
+      LOG.info("Reference {} to retrieve commits no longer exists", reference);
+    }
+    return result;
+  }
+
+  private void handleLiveCommit(
+      LogResponse.LogEntry logEntry,
+      Set<ContentKey> deletedKeys,
+      Map<String, ContentBloomFilter> bloomFilterMap) {
+    if (logEntry.getOperations() != null) {
+      logEntry
+          .getOperations()
+          .forEach(
+              operation -> {
+                if (operation instanceof Operation.Delete) {
```

Comment:
I suspect, this doesn't handle the following case (commits in chronological order):

--> cutoff-time is here
1. create table "foo"
2. rename table "foo" to "bar"
3. create table "foo"
4. drop table "foo"

Feels like content put to the "first foo" before it's being renamed to "bar" will not be added to the bloom filter, right?

Reply:
yes, you are right. I thought of having a distinction between rename table and delete table and skip adding keys to drop table if it is rename. But forgot to implement! 

good catch, I will add it.

Reply:
As I traverse from the head and identify live, from second commit I can check if the keys are deleted and content is not previously marked live using bloom filter (if it is marked live, it cannot be the drop table operations). 

This way I wont add the rename table contents into expired contents.

Reply:
had to use 100% brain for this solution 🤯